src/cyverse_groups_client/core.clj

```diff
@@ -76,3 +78,70 @@
 
   (list-subject-groups [_ user subject]
     "Lists groups that a subject belongs to."))
+
+(defn- build-url [base-url & path-elements]
+  (str (apply curl/url base-url (mapv curl/url-encode path-elements))))
+
+(defn- prepare-opts [opts ks]
+  (remove-vals nil? (select-keys opts ks)))
+
+(defn- folder-name-prefix [environment-name]
+  (format "iplant:de:%s" environment-name))
+
+(deftype CyverseGroupsClient [base-url environment-name]
+  Client
+
+  (get-status [_]
+    (:body (http/get base-url {:as :json})))
+
+  (find-folders [_ user search]
+    (:body (http/get (build-url base-url "folders")
+                     {:query-params {:user user :search search}
+                      :as           :json})))
+
+  (add-folder [self user name description]
+    (add-folder self user name description nil))
+
+  (add-folder [_ user name description display-extension]
+    (:body (http/post (build-url base-url "folders")
+                      {:query-params {:user user}
+                       :form-params  (remove-vals nil? {:name              name
+                                                        :description       description
+                                                        :display_extension display-extension})
+                       :content-type :json
+                       :as           :json})))
+
+  (delete-folder [_ user name]
+    (:body (http/delete (build-url base-url "folders" name)
+                        {:query-params {:user user}
+                         :as           :json})))
+
+  (get-folder [_ user name]
+    (:body (http/get (build-url base-url "folders" name)
+                     {:query-params {:user user}
+                      :as           :json})))
+
+  (update-folder [_ user name updates]
+    (:body (http/put (build-url base-url "folders" name)
+                     {:query-params {:user user}
+                      :form-params  (remove-vals nil? (select-keys updates [:name :description :display_extension]))
```

Comment:
Looks like you could use `prepare-opts` here?
I don't see `prepare-opts` used anywhere else, so it could also just be removed?